bridge-fungible-token/src/utils.sw

```diff
@@ -1,10 +1,158 @@
+    if DECIMALS < LAYER_1_DECIMALS {
+        let amount = ~U256::from(0, 0, 0, val);
+        let factor = ~U256::from(0, 0, 0, 10.pow(LAYER_1_DECIMALS - DECIMALS));
+        let components = amount.multiply(factor).into();
+        compose(components)
+    } else {
+        // Either decimals are the same, or decimals are negative.
+        // TODO: Decide how to handle negative decimals before mainnet.
+        // For now we make no decimal adjustment for either case.
+        compose((0, 0, 0, val))
+    }
+}
+
+/// Make any necessary adjustments to decimals(precision) on the deposited value, and return either a converted u64 or an error if the conversion can't be achieved without overflow or loss of precision.
+pub fn adjust_deposit_decimals(msg_val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    let decomposed = decompose(msg_val);
+    let value = ~U256::from(decomposed.0, decomposed.1, decomposed.2, decomposed.3);
+
+    if LAYER_1_DECIMALS > DECIMALS {
+        let adjustment_factor = ~U256::from(0, 0, 0, 10.pow(LAYER_1_DECIMALS - DECIMALS));
+
+        if value.divide(adjustment_factor).multiply(adjustment_factor) == value
+            && (value.gt(adjustment_factor)
+            || value.eq(adjustment_factor))
+        {
+            let adjusted = value.divide(adjustment_factor);
+            let val_result = adjusted.as_u64();
+            match val_result {
+                Result::Err(e) => {
+                    Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+                },
+                Result::Ok(val) => {
+                    Result::Ok(val)
+                },
+            }
+        } else {
+            Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+        }
+    } else {
+        // Either decimals are the same, or decimals are negative.
+        // TODO: Decide how to handle negative decimals before mainnet.
+        // For now we make no decimal adjustment for either case.
+        let val_result = value.as_u64();
+        match val_result {
+            Result::Err(e) => {
+                Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+            },
+            Result::Ok(val) => {
+                Result::Ok(val)
+            },
+        }
+    }
+}
+
+/// Build a single b256 value from a tuple of 4 u64 values.
+pub fn compose(words: (u64, u64, u64, u64)) -> b256 {
+    let res: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    asm(w0: words.0, w1: words.1, w2: words.2, w3: words.3, result: res) {
+        sw result w0 i0;
+        sw result w1 i1;
+        sw result w2 i2;
+        sw result w3 i3;
+        result: b256
+    }
+}
+
+/// Get a tuple of 4 u64 values from a single b256 value.
+pub fn decompose(val: b256) -> (u64, u64, u64, u64) {
+    let empty_tup = (0u64, 0u64, 0u64, 0u64);
+    asm(r1: __addr_of(val), res1: empty_tup.0, res2: empty_tup.1, res3: empty_tup.2, res4: empty_tup.3, tup: empty_tup) {
```

Comment:
I don't think you need to initialize res1-res4. You can just do `asm(r1: __addr_of(val), res1, res2, res3, res4, tup: empty_tup)`. You could also just use the `MCP` [opcode ](MCP)instead of all the `LW` + `SW` and eliminate the need for those registers. You might even be able to have the compiler cast the b256 as a tuple through some assembly magic like just `asm(r1: __addr_of(val)) { r1: (u64, u64, u64, u64) }`, but I might be wrong